sw/device/tests/dif/dif_uart_test.cc

```diff
@@ -0,0 +1,657 @@
+  size_t bytes_read = std::numeric_limits<size_t>::max();
+  std::vector<uint8_t> receive_bytes(num_bytes);
+  EXPECT_TRUE(dif_uart_bytes_receive(&dif_uart_, num_bytes,
+                                     receive_bytes.data(), &bytes_read));
+
+  EXPECT_EQ(bytes_read, 0);
+  EXPECT_THAT(receive_bytes, Each(Eq(0)));
+}
+
+class BytesSendPolledTest : public UartTest {};
+
+TEST_F(BytesSendPolledTest, UartNull) {
+  EXPECT_FALSE(dif_uart_byte_send_polled(nullptr, 'X'));
+}
+
+TEST_F(BytesSendPolledTest, Success) {
+  // Busy loop 1 iteration (waiting for TX FIFO to free up)
+  EXPECT_READ32(UART_STATUS_REG_OFFSET, {{UART_STATUS_TXFULL, true}});
+  EXPECT_READ32(UART_STATUS_REG_OFFSET, {{UART_STATUS_TXFULL, false}});
+
+  // Set expectations for the byte to be set
+  EXPECT_READ32(UART_STATUS_REG_OFFSET, {{UART_STATUS_TXFULL, false}});
+  EXPECT_WRITE32(UART_WDATA_REG_OFFSET, 'X');
+
+  // Busy loop 1 iteration (waiting for the byte to be sent out by the HW)
+  EXPECT_READ32(UART_STATUS_REG_OFFSET, {{UART_STATUS_TXIDLE, false}});
+  EXPECT_READ32(UART_STATUS_REG_OFFSET, {{UART_STATUS_TXIDLE, true}});
+
+  EXPECT_TRUE(dif_uart_byte_send_polled(&dif_uart_, 'X'));
+}
+
+class BytesReceivePolledTest : public UartTest {};
+
+TEST_F(BytesReceivePolledTest, UartNull) {
+  uint8_t byte;
+  EXPECT_FALSE(dif_uart_byte_receive_polled(nullptr, &byte));
+}
+
+TEST_F(BytesReceivePolledTest, ByteNull) {
+  EXPECT_FALSE(dif_uart_byte_receive_polled(&dif_uart_, nullptr));
+}
+
+TEST_F(BytesReceivePolledTest, Success) {
+  // Busy loop 1 iteration (waiting for RX FIFO to fill up)
+  EXPECT_READ32(UART_STATUS_REG_OFFSET, {{UART_STATUS_RXEMPTY, true}});
+  EXPECT_READ32(UART_STATUS_REG_OFFSET, {{UART_STATUS_RXEMPTY, false}});
+
+  // Set expectations for the byte to be read
+  EXPECT_READ32(UART_STATUS_REG_OFFSET, {{UART_STATUS_RXEMPTY, false}});
+  EXPECT_READ32(UART_RDATA_REG_OFFSET, 'X');
+
+  uint8_t byte = 'Y';
+  EXPECT_TRUE(dif_uart_byte_receive_polled(&dif_uart_, &byte));
+
+  EXPECT_EQ('X', byte);
+}
+
+class IrqStateGetTest : public UartTest {};
+
+TEST_F(IrqStateGetTest, UartNull) {
+  dif_uart_enable_t state;
+  EXPECT_FALSE(
+      dif_uart_irq_state_get(nullptr, kDifUartInterruptTxWatermark, &state));
+}
+
+TEST_F(IrqStateGetTest, StateNull) {
+  EXPECT_FALSE(dif_uart_irq_state_get(&dif_uart_, kDifUartInterruptTxWatermark,
+                                      nullptr));
+}
+
+TEST_F(IrqStateGetTest, Success) {
+  // Get the first IRQ state.
+  EXPECT_READ32(UART_INTR_STATE_REG_OFFSET,
+                {{UART_INTR_STATE_TX_WATERMARK, true}});
+
+  dif_uart_enable_t tx_watermark_state = kDifUartDisable;
+  EXPECT_TRUE(dif_uart_irq_state_get(&dif_uart_, kDifUartInterruptTxWatermark,
+                                     &tx_watermark_state));
+
+  EXPECT_EQ(tx_watermark_state, kDifUartEnable);
+
+  // Get the last IRQ state.
+  EXPECT_READ32(UART_INTR_STATE_REG_OFFSET,
+                {{UART_INTR_STATE_RX_PARITY_ERR, false}});
```

Comment:
Is the intention to return 0 here?

Reply:
Good catch, not anymore I think its from previous iterations of the change. Thanks.